hecuba_py/hecuba/hnumpy.py

```diff
@@ -105,11 +125,56 @@ def load_array(storage_id, name):
                           'write_buffer': config.write_buffer_size,
                           'timestamped_writes': config.timestamped_writes})
         hcache = HNumpyStore(*hcache_params)
-        result = hcache.get_numpy([storage_id])
-        if len(result) == 1:
-            return [result[0], hcache, hcache_params]
+        result = hcache.get_reserved_numpy([storage_id])
+        return [result[0], result[1], hcache, hcache_params]
+
+    def generate_coordinates(self, coordinates):
+        if coordinates is None: return []
+        coord = [coordinates[:, coord] // self._row_elem for coord in
+                 range(len(coordinates[0]))]  # coords divided by number of elem in a row
+        keys = list([coord for coord in it.product(*(range(*r) for r in zip(coord[0], coord[1]+1)))])
+        return keys
+
+    def format_coords(self, coord):
+        if coord == slice(None, None, None) or slice(None, None, None) in coord: return None
+        elif isinstance(coord, slice):
+            coordinates = np.array([coord.start, coord.stop])
         else:
-            raise KeyError
+            coordinates = np.array([[coord.start, coord.stop] for coord in coord])
+        return coordinates
+
+    def slices_match_numpy_shape(self, sliced_coord):
+        if sliced_coord is None: return None
+        for maxshape in self.shape:
+            for slice in sliced_coord:
+                if not (np.min(slice) >= 0 <= np.max(slice) <= maxshape): raise IndexError('Slices are out of numpy bounds')
+
+    def __getitem__(self, sliced_coord):
+        log.info("RETRIEVING NUMPY")
+
+        coo = self.format_coords(sliced_coord)
+        self.slices_match_numpy_shape(coo)
+
+        # coordinates is the union between the loaded coordiantes and the new ones
+        coordinates = list(set(it.chain.from_iterable((self._loaded_coordinates, self.generate_coordinates(coo)))))
+
+        if ((len(coordinates) != len(self._loaded_coordinates)) and not self._numpy_full_loaded) or (not self._numpy_full_loaded and not coordinates):
+            if not coordinates:
+                self._numpy_full_loaded = True
+                coordinates = None
+            self._hcache.get_numpy_from_coordinates([self._storage_id], coordinates, [self.view(np.ndarray)])
+            self._loaded_coordinates = coordinates
+        return super(StorageNumpy, self).__getitem__(sliced_coord)
+
+    def __setitem__(self, sliced_coord, values):
+        log.info("WRITTING NUMPY")
+        numpy = self.view(np.ndarray)
+        numpy[sliced_coord] = values
+        coo = self.format_coords(sliced_coord)
+        coordinates = list(set(it.chain.from_iterable(
+            (self._loaded_coordinates, self.generate_coordinates(coo)))))
+        self._hcache.set_numpy([self._storage_id], [numpy], [self.view(np.ndarray)], coordinates)
```

Comment:
numpy and self.view(np.ndarray) are the same right? Why the same argument twice?